rust/connlib/tunnel/src/lib.rs

```diff
@@ -373,16 +384,22 @@ where
             _ => {}
         }
 
-        if let Poll::Ready(instant) = self.connection_pool_timeout.poll_unpin(cx) {
-            self.node.handle_timeout(instant);
-            if let Some(timeout) = self.node.poll_timeout() {
-                self.connection_pool_timeout = sleep_until(timeout).boxed();
-            }
+        Poll::Pending
+    }
 
-            cx.waker().wake_by_ref();
+    fn poll_timeout(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        if let Some(timeout) = self.node.poll_timeout() {
+            let timeout = tokio::time::Instant::from_std(timeout);
+
+            if timeout != self.timeout.deadline() {
+                self.timeout.as_mut().reset(timeout)
+            }
         }
 
-        Poll::Pending
+        ready!(self.timeout.poll_unpin(cx));
+        self.node.handle_timeout(self.timeout.deadline().into());
+
+        Poll::Ready(())
```

Comment:
Okay so this is exactly like quinn-proto, this tells the driver when to next wake up the state machine for a timeout.